src/async.rs

```diff
@@ -0,0 +1,290 @@
+        }
+    };
+}
+
+macro_rules! with_write_connection {
+    ($con:expr, $f:expr) => {
+        match $con {
+            #[cfg(not(feature = "with-async-unix-sockets"))]
+            ActualConnection::Tcp(con) => {
+                $f(WriteWrapper(con)).map(|(con, value)| (ActualConnection::Tcp(con.0), value))
+            }
+
+            #[cfg(feature = "with-async-unix-sockets")]
+            ActualConnection::Tcp(con) => Either::A(
+                $f(WriteWrapper(con)).map(|(con, value)| (ActualConnection::Tcp(con.0), value)),
+            ),
+            #[cfg(feature = "with-async-unix-sockets")]
+            ActualConnection::Unix(con) => Either::B(
+                $f(WriteWrapper(con))
+                    .map(|(con, value)| (ActualConnection::Unix(con.0), value)),
+            ),
+        }
+    };
+}
+
+impl Connection {
+    pub fn read_response(self) -> RedisFuture<(Self, Value)> {
+        let db = self.db;
+        Box::new(
+            with_connection!(self.con, ::parser::parse_async).then(move |result| {
+                match result {
+                    Ok((con, value)) => Ok((Connection { con: con, db }, value)),
+                    Err(err) => {
+                        // TODO Do we need to shutdown here as we do in the sync version?
+                        Err(err)
+                    }
+                }
+            }),
+        )
+    }
+}
+
+pub fn connect(
+    connection_info: ConnectionInfo,
+    handle: &tokio_reactor::Handle,
+) -> RedisFuture<Connection> {
+    let connection = match *connection_info.addr {
+        ConnectionAddr::Tcp(ref host, port) => {
+            let socket_addr = match (&host[..], port).to_socket_addrs() {
+                Ok(mut socket_addrs) => match socket_addrs.next() {
+                    Some(socket_addr) => socket_addr,
+                    None => {
+                        return Box::new(future::err(RedisError::from((
+                            ErrorKind::InvalidClientConfig,
+                            "No address found for host",
+                        ))))
+                    }
+                },
+                Err(err) => return Box::new(future::err(err.into())),
+            };
+
+            let stream_result = (|| -> io::Result<_> {
+                let builder = match socket_addr {
+                    SocketAddr::V4(_) => TcpBuilder::new_v4()?,
+                    SocketAddr::V6(_) => TcpBuilder::new_v6()?,
+                };
+                Ok(builder.to_tcp_stream()?)
+            })();
+            let stream = match stream_result {
+                Ok(stream) => stream,
+                Err(err) => return Box::new(future::err(err.into())),
+            };
+            Either::A(
+                TcpStream::connect_std(stream, &socket_addr, handle)
+                    .from_err()
+                    .map(|con| ActualConnection::Tcp(BufReader::new(con))),
+            )
+        }
+        #[cfg(feature = "with-async-unix-sockets")]
```

Comment:
It's weird seeing this feature, when it's not available in Cargo.toml.
What do you think about extracting the async-unix-socket code from this PR, but immediately opening one adding it, pending the updated tokio-uds crate?

Reply:
`tokio-uds` actually got a release so I removed the feature and enable `tokio-uds` support with the normal uds feature https://crates.io/crates/tokio-uds